File: examples/python/did_wba_examples/create_did_document.py
"""Minimal example showing how to create a DID WBA document."""

from __future__ import annotations

import argparse
import json
from pathlib import Path

from anp.authentication import create_did_wba_document


def parse_args() -> argparse.Namespace:
    """Parse command-line arguments."""
    parser = argparse.ArgumentParser(
        description="Create a DID WBA document and save generated artifacts.",
    )
    parser.add_argument(
        "--profile",
        choices=("e1", "k1", "plain_legacy"),
        default="e1",
        help="DID profile to generate.",
    )
    parser.add_argument(
        "--hostname",
        default="demo.agent-network",
        help="Hostname used in the DID.",
    )
    return parser.parse_args()


def main() -> None:
    """Create a DID document and persist the generated artifacts."""
    args = parse_args()
    did_document, keys = create_did_wba_document(
        hostname=args.hostname,
        path_segments=["agents", "demo"],
        agent_description_url=f"https://{args.hostname}/agents/demo",
        did_profile=args.profile,
    )

    output_dir = Path(__file__).resolve().parent / "generated" / args.profile
    output_dir.mkdir(parents=True, exist_ok=True)

    did_path = output_dir / "did.json"
    did_path.write_text(json.dumps(did_document, indent=2), encoding="utf-8")
    print(f"DID document saved to {did_path}")

    for fragment, (private_bytes, public_bytes) in keys.items():
        private_path = output_dir / f"{fragment}_private.pem"
        public_path = output_dir / f"{fragment}_public.pem"
        private_path.write_bytes(private_bytes)
        public_path.write_bytes(public_bytes)
        print(
            "Registered verification method",
            fragment,
            "→ private key:",
            private_path.name,
            "public key:",
            public_path.name,
        )

    print(f"Generated DID identifier: {did_document['id']}")
    proof = did_document.get("proof", {})
    if proof:
        print(
            "Generated proof profile:",
            proof.get("type"),
            proof.get("cryptosuite", "<legacy>"),
        )


if __name__ == "__main__":
    main()

File: rust/examples/create_did_document.rs
use std::fs;
use std::path::PathBuf;

use anp::authentication::{create_did_wba_document, DidDocumentOptions, DidProfile};

fn main() {
    let mut profile = DidProfile::E1;
    let mut hostname = "demo.agent-network".to_string();

    let mut args = std::env::args().skip(1);
    while let Some(arg) = args.next() {
        match arg.as_str() {
            "--profile" => {
                if let Some(value) = args.next() {
                    profile = DidProfile::from_str(&value)
                        .expect("profile must be one of: e1, k1, plain_legacy");
                }
            }
            "--hostname" => {
                hostname = args.next().expect("hostname value is required");
            }
            _ => {}
        }
    }

    let bundle = create_did_wba_document(
        &hostname,
        DidDocumentOptions::default()
            .with_profile(profile)
            .with_path_segments(["agents", "demo"])
            .with_agent_description_url(format!("https://{hostname}/agents/demo")),
    )
    .expect("DID creation should succeed");

    let output_dir = PathBuf::from("examples/generated").join(profile.as_str());
    fs::create_dir_all(&output_dir).expect("output directory should be created");

    let did_path = output_dir.join("did.json");
    fs::write(
        &did_path,
        serde_json::to_vec_pretty(&bundle.did_document).expect("DID document should serialize"),
    )
    .expect("DID document should be written");
    println!("DID document saved to {}", did_path.display());

    for (fragment, key_pair) in &bundle.keys {
        let private_path = output_dir.join(format!("{}_private.pem", fragment));
        let public_path = output_dir.join(format!("{}_public.pem", fragment));
        fs::write(&private_path, &key_pair.private_key_pem).expect("private key should be written");
        fs::write(&public_path, &key_pair.public_key_pem).expect("public key should be written");
        println!(
            "Registered verification method {} -> private key: {} public key: {}",
            fragment,
            private_path.file_name().unwrap().to_string_lossy(),
            public_path.file_name().unwrap().to_string_lossy(),
        );
    }

    println!(
        "Generated DID identifier: {}",
        bundle.did().unwrap_or("<unknown>")
    );
    if let Some(proof) = bundle.did_document.get("proof") {
        println!(
            "Generated proof profile: {} {}",
            proof
                .get("type")
                .and_then(serde_json::Value::as_str)
                .unwrap_or("<unknown>"),
            proof
                .get("cryptosuite")
                .and_then(serde_json::Value::as_str)
                .unwrap_or("<legacy>"),
        );
    }
}
